db/migrations/dedupe_github_repo_owner.py

```diff
@@ -0,0 +1,61 @@
+"""Remove case-sensitive duplicates in GitHub repository tables"""
+
+import rethinkdb as r
+
+import db.plugins
+import db.util
+
+r_conn = db.util.r_conn
+
+
+if __name__ == '__main__':
+    table = 'plugin_github_repos'
+
+    print 'Removing duplicate rows in %s' % table
+
+    updated = 0
+    deleted = 0
+
+    query = r.table(table)
+
+    # Group by the normalized GitHub path
+    query = query.group([
+        r.row['owner'].downcase(),
+        r.row['repo_name'].downcase()])
+
+    # Get the most recently scraped row in each group first
+    query = query.order_by(r.desc('last_scaped_at'))
+
+    grouped_repos = query.run(r_conn())
+
+    for owner_repo, repos in grouped_repos.iteritems():
+
+        print '\nRepo with GitHub path %s occurs %s times' % (
+                owner_repo,
+                len(repos))
+
+        # Use the most recently scraped row as the canonical row
+        canonical = repos.pop()
+
+        assert canonical
```

Comment:
It's intended as a sanity check. Basically: "if canonical is falsey here, something has gone wrong and we should just crash out"

It also doubles as a sort of comment: "from this point on, we're operating under the assumption that we have a canonical item"

--
Jordan Eldredge

Please excuse any typos or terseness; this email was sent from a mobile device. 

> On Mar 5, 2016, at 3:20 PM, Jon Banafato <notifications@github.com> wrote:
> 
> In db/migrations/dedupe_github_repo_owner.py:
> 
> > +
> > +    # Get the most recently scraped row in each group first
> > +    query = query.order_by(r.desc('last_scaped_at'))
> > +
> > +    grouped_repos = query.run(r_conn())
> > +
> > +    for owner_repo, repos in grouped_repos.iteritems():
> > +
> > +        print '\nRepo with GitHub path %s occurs %s times' % (
> > +                owner_repo,
> > +                len(repos))
> > +
> > +        # Use the most recently scraped row as the canonical row
> > +        canonical = repos.pop()
> > +
> > +        assert canonical
> What's the purpose of this assert statement here? As written, if canonical is falsey in an iteration of the loop, no later iterations will be run. Would this be better as something like the following?
> 
> if not canonical:
>     continue
> —
> Reply to this email directly or view it on GitHub.
>